### PR TITLE
Update lib versions

### PR DESCRIPTION
**Pull request by Ennosigaeon**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](doc/contribution-guidelines.md#commit-message-format) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [x] Docs have been added / updated (for bug fixes / features) ?

Most referenced libraries are quite outdated. This PR upgrades most of them to the latest compatible version. I also decided to replace joda DateTime with java.time.ZonedDataTime as joda should not be used for Java 8+ projects (see https://www.joda.org/joda-time/). 

I tried to check to the best of my abilities that the behavior did not change at all. Yet, I surely missed some corner cases. It would be great if you can also give it a test.

**Other information**:
derby, h4 and flyway are still quite outdated but updating them are breaking stuff. Therefore, I decided to bump them to the latest non-breaking version.

---
This change is [Reviewable](https://reviewable.io/reviews/spark-jobserver/spark-jobserver/1356)

### CODE REVIEW

File: job-server/src/main/scala/spark/jobserver/io/DataFileDAO.scala [hunk 1]
```diff
@@ -134,16 +135,17 @@ class DataFileDAO(config: Config) {
...
   private def addFile(aName: String) {
     files += aName
   }
 
   def listFiles: Set[String] = files.toSet
 
-  private def createFileName(aName: String, uploadTime: DateTime): String =
-    aName + "-" + uploadTime.toString().replace(':', '_')
+  private def createFileName(aName: String, uploadTime: ZonedDateTime): String =
+    aName + "-" + DateTimeFormatter.ISO_INSTANT.format(uploadTime).replace(':', '_')
```
Comment (velvia) [inline job-server/src/main/scala/spark/jobserver/io/DataFileDAO.scala:148]: Hm, is this time format the same?  (Wouldn't toString still work the same way?)

**Review by velvia — APPROVED** [review]
@Ennosigaeon Thanks very much, changes look mostly mechanical and fine to me.
Need to figure out why the tests are failing.

**Comment by Ennosigaeon** [comment]
@velvia If I have seen it correctly, only spark.jobserver.AkkaClusterSupervisorActorSpec failed in Travis. I have run the tests locally and all tests passed.

**Comment by valan4ik** [comment]
@Ennosigaeon yes, I also saw that it's only `AkkaClusterSupervisorActorSpec`, but somehow can't restart the tests on Jenkins, get "something went wrong" error from Travis.
Maybe you could close and open PR again?
This is another trick to restart the tests.

File: job-server/src/main/scala/spark/jobserver/io/DataFileDAO.scala (hunk 1 above)
Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/io/DataFileDAO.scala:148]: Good catch. I only checked UTC and for other time zones file names were not identical. Fixed it now. The normal `toString` changed with Java time producing something like `2021-05-06T10_44_20.632275+02_00[Europe/Berlin]`

**Comment by Ennosigaeon** [comment]
Ok, several retires later and all tests are green...

**Comment by velvia** [comment]
@valan4ik I think we can merge this right?

File: job-server/src/main/scala/spark/jobserver/util/JsonProtocols.scala [hunk 2]
```diff
@@ -1,16 +1,17 @@
...
 import spark.jobserver.io.{BinaryInfo, BinaryType, ContextInfo, JobInfo, JobStatus}
 import spray.json._
 
+import java.time.format.DateTimeFormatter
+import java.time.{ZoneId, ZonedDateTime}
+
 object JsonProtocols extends DefaultJsonProtocol {
 
-  val DATE_PATTERN = "yyyy-MM-dd HH-mm-ss SS Z"
+  val DATE_PATTERN = "yyyy-MM-dd'T'HH:mm:ss.SSSX"
```
Comment (valan4ik) [inline job-server/src/main/scala/spark/jobserver/util/JsonProtocols.scala:14]: Will this change break reading existing data from the database?
If I already have some contexts/jobs data in DB, will they be parsed correctly by the new Jobserver?
There is no such test in this change.

Comment (Ennosigaeon) [reply job-server/src/main/scala/spark/jobserver/util/JsonProtocols.scala]: Ok, two things:
1. The pattern was actually wrong, I have corrected it and added two test cases to check it. The expected strings are extracted from the Readme but I also verified that the old behaviour of SJS also conforms to this pattern.
2. I am not sure where this particular pattern was actually used. I was not able to find it in any user-exposed data types/interfaces. They all use the new pattern (which is identical to the default of JodaTime.toString()).

Comment (valan4ik) [reply job-server/src/main/scala/spark/jobserver/util/JsonProtocols.scala]: This JSON protocols are at least used in Zookeeper DAO.
I now made a simple test by starting a Zookeeper docker container and configuring DAO as described in https://github.com/spark-jobserver/spark-jobserver/blob/master/doc/dao-setup.md#hdfs--zookeeper

I created a context and could see it in Zookeeper and in CLI:
```
[zk: localhost:2181(CONNECTED) 13] get /jobserver/db/contexts/94a9e08a-4edc-4843-b377-e4eb639b278e
{"actorAddress":null,"config":"{\"context\":{\"name\":\"test\"},\"context-factory\":\"spark.jobserver.context.DefaultSparkContextFactory\",\"context-init-timeout\":\"60s\",\"forked-jvm-init-timeout\":\"30s\",\"is-adhoc\":\"false\",\"launcher\":{\"spark\":{\"driver\":{\"memory\":\"1g\"}}},\"memory-per-node\":\"512m\",\"num-cpu-cores\":2,\"passthrough\":{\"spark\":{\"driver\":{\"allowMultipleContexts\":true}}},\"python\":{\"executable\":\"python3\",\"paths\":[]},\"spark\":{\"session\":{\"hive\":{\"enabled\":true}}},\"streaming\":{\"batch_interval\":1000,\"stopGracefully\":true,\"stopSparkContext\":true}}","endTime":"2021-05-11 11-56-24 531 +0200","error":"Context failed to connect back within initialization time","id":"94a9e08a-4edc-4843-b377-e4eb639b278e","name":"test","startTime":"2021-05-11 11-56-24 532 +0200","state":"ERROR"}
```
```
curl localhost:8090/contexts/test
{"name":"test","startTime":"2021-05-11T11:56:24.532+02:00","state":"ERROR","endTime":"2021-05-11T11:56:24.531+02:00","id":"94a9e08a-4edc-4843-b377-e4eb639b278e"}
```


After starting Jobserver with current change, context metadata was not anymore readable:
```
[2021-05-11 12:01:48,911] ERROR o.zookeeper.ZookeeperUtils [] [akka://JobServer/user/dao-manager] - Text '2021-05-11 11-56-24 531 +0200' could not be parsed
at index 10 : java.time.format.DateTimeParseException: Text '2021-05-11 11-56-24 531 +0200' could not be parsed at index 10
        at java.time.format.DateTimeFormatter.parseResolved0(DateTimeFormatter.java:1949) 
```

I didn't do such test for SQL DAO yet, but I think it also should be checked.

**Comment by valan4ik** [comment]
@Ennosigaeon thanks for making the change and reacting on feedback despite delayed reviews.
The change LGTM, it looks like the failing tests are unrelated.


Just in case you have time, I think one of the very flaky tests was introduced with callback PR:
``` 
[info] - should start job, return JobStarted (async) and invoke callback *** FAILED *** (207 milliseconds)
[info]   0 was not equal to 1 (JobManagerActorSpec.scala:487)
[info]   org.scalatest.exceptions.TestFailedException:
``` 
Looks like some concurrent requests issue, maybe JobInfo takes a bit more time to get updated

**Review by valan4ik — APPROVED** [review]

**Comment by Ennosigaeon** [comment]
@valan4ik Sure, no problem. I am in no hurry to get these PRs closed.
I will take a look at the failing test next week.

**Comment by valan4ik** [comment]
@Ennosigaeon , could you please rebase on master branch? 
There is a conflict now and I also hope we can get green tests after recent improvements.

**Comment by Ennosigaeon** [comment]
Ok, let's hope that it wasn't pure coincidence that the tests passed this time.
